client/src/test/java/io/pravega/client/stream/impl/StreamSegmentsTest.java

```diff
@@ -124,6 +125,99 @@ public void testRangeReplacementMerge() {
+        newRange.put(new SegmentWithRange(getSegment(7, 3), 0.125, 0.25), ImmutableList.of(computeSegmentId(2, 1)));
+        streamSegments = streamSegments.withReplacementRange(new StreamSegmentsWithPredecessors(newRange, ""));
+
+        assertEquals(getSegment(6, 3), streamSegments.getSegmentForKey(0.1));
+        assertEquals(getSegment(7, 3), streamSegments.getSegmentForKey(0.24));
+        assertEquals(getSegment(3, 1), streamSegments.getSegmentForKey(0.4));
+        assertEquals(getSegment(1, 0), streamSegments.getSegmentForKey(0.7));
+
+        // simulate fetch successors for segment 3.
+        newRange = new HashMap<>();
+        newRange.put(new SegmentWithRange(getSegment(8, 4), 0.25, 0.75),
+                ImmutableList.of(computeSegmentId(3, 1), computeSegmentId(4, 2)));
+        streamSegments = streamSegments.withReplacementRange(new StreamSegmentsWithPredecessors(newRange, ""));
+
+        assertEquals(getSegment(6, 3), streamSegments.getSegmentForKey(0.1));
+        assertEquals(getSegment(7, 3), streamSegments.getSegmentForKey(0.24));
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.4));
+        //this should not return segment 8 as segment 4 is never updated.
+        assertEquals(getSegment(1, 0), streamSegments.getSegmentForKey(0.7));
+
+        // simulate fetch successors for segment 1.
+        newRange = new HashMap<>();
+        newRange.put(new SegmentWithRange(getSegment(4, 2), 0.5, 0.75), ImmutableList.of(computeSegmentId(1, 0)));
+        newRange.put(new SegmentWithRange(getSegment(5, 2), 0.75, 1.0), ImmutableList.of(computeSegmentId(1, 0)));
+        streamSegments = streamSegments.withReplacementRange(new StreamSegmentsWithPredecessors(newRange, ""));
+
+        assertEquals(getSegment(6, 3), streamSegments.getSegmentForKey(0.1));
+        assertEquals(getSegment(7, 3), streamSegments.getSegmentForKey(0.24));
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.4));
+        assertEquals(getSegment(4, 2), streamSegments.getSegmentForKey(0.7)); //this should now return segment 4 and not segment 1.
+        assertEquals(getSegment(5, 2), streamSegments.getSegmentForKey(0.9)); //this should now return segment 5 and not segment 1.
+
+        // simulate fetch successors for segment 4.
+        newRange = new HashMap<>();
+        newRange.put(new SegmentWithRange(getSegment(8, 4), 0.25, 0.75),
+                ImmutableList.of(computeSegmentId(3, 1), computeSegmentId(4, 2)));
+        streamSegments = streamSegments.withReplacementRange(new StreamSegmentsWithPredecessors(newRange, ""));
+
+        assertEquals(getSegment(6, 3), streamSegments.getSegmentForKey(0.1));
+        assertEquals(getSegment(7, 3), streamSegments.getSegmentForKey(0.24));
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.4));
+        // this should now return segment 8 and not segment 4
+        assertEquals(getSegment(8, 4), streamSegments.getSegmentForKey(0.7));
```

Comment:
At the end of the test 8 is actually in the map twice with two different ranges. That might not be an issue, but it seems like a lurking complexity that could become a bug in the future if the code changed without thinking about that.

Reply:
With the current code there will be two entries i.e
Key : 0.5 Value: `8.#epoch.4`
Key: 0.75 Value: `8.#epoch.4`

Reply:
then we should include only one range

Reply:
Yes, I worry about that when 8 is replaced. It is very easy to assume that it won't be in the list twice...

Reply:
updated the code to handle compaction. Will add a test case for the same.